### PR TITLE
In case of an error, we didn't disconnect

### PR DESCRIPTION
**Pull request by marco44**
the session was left idle in transaction aborted

### CODE REVIEW

File: powa_collector/powa_worker.py [hunk 1]
```diff
@@ -400,10 +400,10 @@ def __worker_main(self):
                     self.logger.error("Error during snapshot: %s" % e)
                     if (self.__repo_conn is None
                             or self.__repo_conn.closed > 0):
-                        self.__repo_conn = None
```
Comment (rjuju) [inline powa_collector/powa_worker.py:403]: mmm, actually the first test is a bit cryptic.  Should we change the test for "is not None and..."?

Then, according to psycopg2 doc, if connection.closed > 0, then it's closed or broken.  I'm assuming that "broken" actually means that there's still a backend opened?

File: powa_collector/powa_worker.py [hunk 2]
```diff
@@ -400,10 +400,10 @@ def __worker_main(self):
...
+                        self.__disconnect_repo()
                     if (self.__remote_conn is None
                             or self.__remote_conn.closed > 0):
-                        self.__remote_conn = None
+                        self.__disconnect_repo()
```
Comment (rjuju) [inline powa_collector/powa_worker.py:406]: This should handle the remote connection :)

**Review by rjuju — CHANGES_REQUESTED** [review]

File: powa_collector/powa_worker.py (hunk 1 above)
Comment (marco44) [reply powa_collector/powa_worker.py:403]: I think this test is good. What do you think is a problem ?

File: powa_collector/powa_worker.py (hunk 2 above)
Comment (marco44) [reply powa_collector/powa_worker.py:406]: It's not closing the correct one?

File: powa_collector/powa_worker.py (hunk 1 above)
Comment (rjuju) [reply powa_collector/powa_worker.py:403]: If the repository connection is None, then by definition there's no cleanup to do (and __disconnect_repo() will check that and do nothing in that case).  Here we just need to check that there's an object, so we don't try to access an unexisting property.

File: powa_collector/powa_worker.py (hunk 2 above)
Comment (rjuju) [reply powa_collector/powa_worker.py:406]: Yes, __disconnect_repo() only closes the repository connection.  There's no function to close the remote connection as it's supposed to be a persistent connection.

**Comment by rjuju** [comment]
@marco44 any news on that?

**Comment by rjuju** [comment]
So, coming back on that PR, do we really want to disconnect the repository and remote connections in case of error, or should we try to perform a rollback only?

The full disconnect approach seems safer, as we won't be able to miss an aborted transaction anywhere or something, but I'm wondering if we can end up with a storm of disconnect / connect.  If you have enough remote server and there's any (temporary) issue on the repository server you can end up with all threads trying to reconnect forever.  With a 1 min frequency and a hundred remote servers, that would be more than 1 cycle per second on average.

**Comment by marco44** [comment]
Yeah, we can try a simple rollback, depending on the psycopg2 exception. I'd say if we get a "DatabaseError", just rollback. Else it's something more serious, and a rollback won't cut it...

**Comment by rjuju** [comment]
So you mean closing both connections for any exception except DatabaseError?  I'm fine with it.

While at it we could also probably improve `__connect()`.  For now it's only trying to reconnect if the connection were explicitly set to None, but there are probably unhandled corner cases where the connection could end up being set and with `closed` > 0.  We should try to detect that first and set the underlying connection to None?

**Comment by marco44** [comment]
Yeah, we may even simplify everything by just calling __connect() unconditionally and have it reconnect if needed. We could just check if the connection is either None, or it's closed attribute is true and reconnect...

File: powa_collector/powa_worker.py (hunk 2 above)
Comment (marco44) [reply powa_collector/powa_worker.py]: Fixed this silly error :)

**Review by rjuju — APPROVED** [review]

**Pull request merged**